azext_iot/common/central_functions.py

```diff
@@ -0,0 +1,149 @@
+# coding=utf-8
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+from knack.util import CLIError
+from msrestazure.azure_exceptions import CloudError
+from azext_iot.common._azure import _get_aad_token
+
+
+def _get_dps_sas_auth_header(
+    scope_id, device_id, key,
+):
+    import time
+    import base64
+    import hmac
+    import hashlib
+    import urllib
+
+    sr = "{}%2Fregistrations%2F{}".format(scope_id, device_id)
+    expires = int(time.time() + 21600)
+    registration_id = f"{sr}\n{str(expires)}"
+    secret = base64.b64decode(key)
+    signature = base64.b64encode(
+        hmac.new(
+            secret, msg=registration_id.encode("utf8"), digestmod=hashlib.sha256
+        ).digest()
+    )
+    quote_signature = urllib.parse.quote(signature, "~()*!.'")
+    token = f"SharedAccessSignature sr={sr}&sig={quote_signature}&se={str(expires)}&skn=registration"
+    return token
+
+
+def validate_response_payload(response):
+    if response.status_code in [200, 201]:
+        return response.json()
+    exp = CloudError(response)
+    exp.request_id = response.headers.get("x-ms-request-id")
+    raise exp
+
+
+def get_iot_central_device_list(app_id, token):
+    import requests
+
+    host = get_app_host(app_id, token)["host"]
+    url = "https://{}/api/preview/devices".format(host)
+    response = requests.get(url, headers={"Authorization": "Bearer {}".format(token)})
+    return validate_response_payload(response)
+
+
+def get_app_host(app_id, token):
+    import requests
+
+    url = "https://apps.azureiotcentral.com/api/preview/applications/{}".format(app_id)
+    response = requests.get(url, headers={"Authorization": "Bearer {}".format(token)})
+    return validate_response_payload(response)
+
+
+def show_iot_central_provisioning_information(cmd, app_id, device_id):
+    aad_token = _get_aad_token(cmd, resource="https://apps.azureiotcentral.com")[
+        "accessToken"
+    ]
+    if device_id is None:
+        device_list = get_iot_central_device_list(app_id, aad_token)
+        print("Device Provisioning Summary:")
+        for item in device_list.get("value"):
```

Comment:
We will rarely print the result in the CLI. Please construct a JSON like object (with Python elements of course) and return it. You can use json.dumps() to verify you are building a JSON compatible result.